File: Classes/PSResizing.m
//
//  PSResizing.m
//  PocketSword
//
//  Created by Nic Carter on 1/10/10.
//  Copyright 2010 CrossWire Bible Society. All rights reserved.
//
#import <sys/xattr.h>

#import "PSResizing.h"
#import "globals.h"

#define TOP_BAR_LANDSCAPE_HEIGHT 32.0
#define TOP_BAR_PORTRAIT_HEIGHT 44.0
#define BOTTOM_BAR_LANDSCAPE_HEIGHT 32.0
#define BOTTOM_BAR_PORTRAIT_HEIGHT 44.0
//#define TAB_BAR_PORTRAIT_HEIGHT 49.0
//#define TAB_BAR_LANDSCAPE_HEIGHT 35.0

@implementation PSResizing

+(void)resizeViewsOnAppearWithTabBarController:(UITabBarController*)tabBarController topBar:(UIView*)topBar mainView:(UIView*)mainView useStatusBar:(BOOL)useStatusBar {
	[PSResizing resizeViewsOnAppearWithTabBarController:tabBarController topBar:topBar mainView:mainView bottomBar:nil useStatusBar:useStatusBar];
}

+(void)resizeViewsOnAppearWithTabBarController:(UITabBarController*)tabBarController topBar:(UIView*)topBar mainView:(UIView*)mainView bottomBar:(UIView*)bottomBar useStatusBar:(BOOL)useStatusBar {
	CGSize screen = [PSResizing mainScreenBounds].size;
	CGFloat topBarHeight, bottomBarHeight, viewHeight, width;//, tabBarY;
	CGFloat tabBarHeight = (tabBarController) ? tabBarController.tabBar.frame.size.height : 0.0;
	BOOL redrawInNewFrames = NO;
	
	UIInterfaceOrientation interfaceOrientation = [PSResizing currentInterfaceOrientation];
	CGFloat statusBarSize = [PSResizing statusBarHeight];
	if(interfaceOrientation == UIInterfaceOrientationLandscapeLeft || interfaceOrientation == UIInterfaceOrientationLandscapeRight) {
		//DLog(@"landscape");
		redrawInNewFrames = YES;
		bottomBarHeight = (bottomBar) ? BOTTOM_BAR_LANDSCAPE_HEIGHT : 0.0;
		width = screen.height;
		topBarHeight = ([PSResizing iPad]) ? TOP_BAR_PORTRAIT_HEIGHT : TOP_BAR_LANDSCAPE_HEIGHT;
		viewHeight = screen.width - topBarHeight - tabBarHeight - bottomBarHeight;
		if(useStatusBar) {
			viewHeight -= statusBarSize;
		}
	} else if(interfaceOrientation == UIInterfaceOrientationPortrait || interfaceOrientation == UIInterfaceOrientationPortraitUpsideDown) {
		//DLog(@"portrait");
		redrawInNewFrames = YES;
		bottomBarHeight = (bottomBar) ? BOTTOM_BAR_PORTRAIT_HEIGHT : 0.0;
		width = screen.width;
		topBarHeight = TOP_BAR_PORTRAIT_HEIGHT;
		viewHeight = screen.height - topBarHeight - tabBarHeight - bottomBarHeight;
		if(useStatusBar) {
			viewHeight -= statusBarSize;
		}
	}
	if(redrawInNewFrames) {
		topBar.frame = CGRectMake(0.0, 0.0, width, topBarHeight);
		[topBar layoutIfNeeded];
		[topBar setNeedsDisplay];
		if(bottomBar) {
			bottomBar.frame = CGRectMake(0.0, (topBarHeight+viewHeight), width, bottomBarHeight);
			[bottomBar layoutIfNeeded];
			[bottomBar setNeedsDisplay];
		}
		mainView.frame = CGRectMake(0.0, topBarHeight, width, viewHeight);
	}
}

+(void)resizeViewsOnRotateWithTabBarController:(UITabBarController*)tabBarController topBar:(UIView*)topBar mainView:(UIView*)mainView fromOrientation:(UIInterfaceOrientation)fromInterfaceOrientation toOrientation:(UIInterfaceOrientation)toInterfaceOrientation {
	[PSResizing resizeViewsOnRotateWithTabBarController:tabBarController topBar:topBar mainView:mainView bottomBar:nil fromOrientation:fromInterfaceOrientation toOrientation:toInterfaceOrientation];
}

+(void)resizeViewsOnRotateWithTabBarController:(UITabBarController*)tabBarController topBar:(UIView*)topBar mainView:(UIView*)mainView bottomBar:(UIView*)bottomBar fromOrientation:(UIInterfaceOrientation)fromInterfaceOrientation toOrientation:(UIInterfaceOrientation)toInterfaceOrientation {

	if([PSResizing iPad]) {
		return;
	}
	CGSize screen = [PSResizing mainScreenBounds].size;
	CGFloat topBarHeight, bottomBarHeight, viewHeight, width, bottomBarY;
	CGFloat tabBarHeight = (tabBarController) ? tabBarController.tabBar.frame.size.height : 0.0;
	BOOL redrawInNewFrames = NO;
	CGFloat statusBarSize = [PSResizing statusBarHeight];
	if((toInterfaceOrientation == UIInterfaceOrientationLandscapeLeft || toInterfaceOrientation == UIInterfaceOrientationLandscapeRight) && !(fromInterfaceOrientation == UIInterfaceOrientationLandscapeLeft || fromInterfaceOrientation == UIInterfaceOrientationLandscapeRight)) {
		width = screen.width;
		bottomBarHeight = (bottomBar) ? BOTTOM_BAR_LANDSCAPE_HEIGHT : 0.0;
		bottomBarY = screen.height - statusBarSize - (bottomBarHeight / 2.0);
		topBarHeight = TOP_BAR_LANDSCAPE_HEIGHT;
		viewHeight = screen.height - topBarHeight - tabBarHeight - bottomBarHeight - statusBarSize;
		redrawInNewFrames = YES;
	} else if((toInterfaceOrientation == UIInterfaceOrientationPortrait || toInterfaceOrientation == UIInterfaceOrientationPortraitUpsideDown) && !(fromInterfaceOrientation == UIInterfaceOrientationPortrait || fromInterfaceOrientation == UIInterfaceOrientationPortraitUpsideDown)) {
		width = screen.height;
		bottomBarHeight = (bottomBar) ? BOTTOM_BAR_PORTRAIT_HEIGHT : 0.0;
		bottomBarY = screen.width - statusBarSize - (bottomBarHeight / 2.0);
		topBarHeight = TOP_BAR_PORTRAIT_HEIGHT;
		viewHeight = screen.width - topBarHeight - tabBarHeight - bottomBarHeight - statusBarSize;
		redrawInNewFrames = YES;
	}
	if(redrawInNewFrames) {
		//NSLog(@"bottomBarY = %f, bottomBarHeight = %f, bottomBarYcentre = %f, topBarHeight = %f, width = %f, viewHeight = %f", bottomBarY, bottomBarHeight, (bottomBarY + (bottomBarHeight / 2.0)), topBarHeight, width, viewHeight);
		topBar.frame = CGRectMake(0.0, 0.0, width, topBarHeight);
		mainView.center = CGPointMake((width / 2.0), ((viewHeight / 2.0) + topBarHeight));
		mainView.bounds = CGRectMake(0.0, 0.0, width, viewHeight);
		if(bottomBar) {
			bottomBar.center = CGPointMake((width / 2.0), bottomBarY);
			bottomBar.bounds = CGRectMake(0.0, 0.0, width, bottomBarHeight);
		}
	}
}

+ (CGRect)getOrientationRect:(UIInterfaceOrientation)interfaceOrientation {
	CGSize screen = [PSResizing mainScreenBounds].size;
	return CGRectMake(0.0, 0.0, screen.width, screen.height);
}

+ (UIWindowScene *)currentWindowScene {
	for (UIScene *scene in [UIApplication sharedApplication].connectedScenes) {
		if ([scene isKindOfClass:[UIWindowScene class]]) {
			return (UIWindowScene *)scene;
		}
	}
	return nil;
}

+ (CGRect)mainScreenBounds {
	UIWindowScene *windowScene = [PSResizing currentWindowScene];
	if (windowScene) {
		return windowScene.screen.bounds;
	}
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
	return [UIScreen mainScreen].bounds;
#pragma clang diagnostic pop
}

+ (CGFloat)mainScreenScale {
	UIWindowScene *windowScene = [PSResizing currentWindowScene];
	if (windowScene) {
		return windowScene.screen.scale;
	}
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
	return [UIScreen mainScreen].scale;
#pragma clang diagnostic pop
}

+ (UIInterfaceOrientation)currentInterfaceOrientation {
	UIWindowScene *windowScene = [PSResizing currentWindowScene];
	if (windowScene) {
		return windowScene.effectiveGeometry.interfaceOrientation;
	}
	return UIInterfaceOrientationPortrait;
}

+ (CGFloat)statusBarHeight {
	UIWindowScene *windowScene = [PSResizing currentWindowScene];
	if (windowScene && windowScene.statusBarManager) {
		CGRect statusBarFrame = windowScene.statusBarManager.statusBarFrame;
		return statusBarFrame.size.height;
	}
	return 0.0;
}

+ (UIInterfaceOrientationMask)supportedInterfaceOrientations {
	NSInteger rotationLockPosition = [[NSUserDefaults standardUserDefaults] integerForKey:ROTATION_LOCK_POSITION];
	switch (rotationLockPosition) {
		case RotationEnabled:
		{
			if([PSResizing iPad]) {
				return UIInterfaceOrientationMaskAll;
			} else {
				return UIInterfaceOrientationMaskAllButUpsideDown;
			}
		}
			break;
		case RotationLockedInLandscape:
		{
			return UIInterfaceOrientationMaskLandscape;
		}
			break;
		case RotationLockedInPortrait:
		{
			if([PSResizing iPad]) {
				return (UIInterfaceOrientationMaskPortrait | UIInterfaceOrientationMaskPortraitUpsideDown);
			} else {
				return UIInterfaceOrientationMaskPortrait;
			}
		}
			break;
		default:
			return UIInterfaceOrientationMaskAll;
	}
}

+ (BOOL)iPad {
	return ([[UIDevice currentDevice] userInterfaceIdiom] != UIUserInterfaceIdiomPhone);
}

+ (BOOL)addSkipBackupAttributeToItemAtPath:(NSString *)path {
	// make sure we're not backing up this folder!
	DLog(@"Don't Backup:\n---\n%@\n---\n", path);
	
//	if (SYSTEM_VERSION_GREATER_THAN_OR_EQUAL_TO(@"5.1") && &NSURLIsExcludedFromBackupKey) {
//	if(&NSURLIsExcludedFromBackupKey) {
		// iOS 5.1 and later:
		NSURL *URL = [NSURL fileURLWithPath:path isDirectory:YES];
		
		assert([[NSFileManager defaultManager] fileExistsAtPath: [URL path]]);
		
		NSError *error = nil;
		BOOL success = [URL setResourceValue: [NSNumber numberWithBool: YES]
									  forKey: NSURLIsExcludedFromBackupKey error: &error];
		if(!success){
			ALog(@"Error excluding %@ from backup %@", [URL lastPathComponent], error);
		}
		return success;
//	}
//	else {
//		// iOS 5.0.1 and earlier method, as this is backwards compatible with iOS 3.0 :P
//		const char* filePath = [path fileSystemRepresentation];
//		
//		const char* attrName = "com.apple.MobileBackup";
//		u_int8_t attrValue = 1;
//		
//		int result = setxattr(filePath, attrName, &attrValue, sizeof(attrValue), 0, 0);
//		return result == 0;
//	}
//	return YES;
}

@end

File: Classes/PSResizing.swift
//
//  PSResizing.swift
//  PocketSword
//
//  Created by Nic Carter on 1/10/10.
//  Copyright 2010 CrossWire Bible Society. All rights reserved.
//
//  Swift port (migration step 0d) of the former Classes/PSResizing.{h,m}.
//  Pure UIKit geometry / orientation / window helpers, no app-internal deps.
//  The @objc surface reproduces the original Obj-C public API 1:1 so that the
//  26 Obj-C bodies and the one header importer bind unchanged.
//

import UIKit

private let TOP_BAR_LANDSCAPE_HEIGHT: CGFloat = 32.0
private let TOP_BAR_PORTRAIT_HEIGHT: CGFloat = 44.0
private let BOTTOM_BAR_LANDSCAPE_HEIGHT: CGFloat = 32.0
private let BOTTOM_BAR_PORTRAIT_HEIGHT: CGFloat = 44.0

// Rotation-lock ordinals. Mirror of the `RotationPosition` NS_ENUM in
// globals.h (RotationEnabled = 0, RotationLockedInPortrait = 1,
// RotationLockedInLandscape = 2). That enum is NOT in the Swift bridging
// header (per 0b, globals.h is mirrored into AppConstants.swift rather than
// bridged), so we reproduce the load-bearing ordinals here. Keep byte-identical
// to globals.h until that enum is fully owned by Swift.
private let kRotationEnabled = 0
private let kRotationLockedInPortrait = 1
private let kRotationLockedInLandscape = 2

@objc(PSResizing)
final class PSResizing: NSObject {

    @objc(resizeViewsOnAppearWithTabBarController:topBar:mainView:useStatusBar:)
    class func resizeViewsOnAppear(withTabBarController tabBarController: UITabBarController?,
                                   topBar: UIView?,
                                   mainView: UIView?,
                                   useStatusBar: Bool) {
        PSResizing.resizeViewsOnAppear(withTabBarController: tabBarController,
                                       topBar: topBar,
                                       mainView: mainView,
                                       bottomBar: nil,
                                       useStatusBar: useStatusBar)
    }

    @objc(resizeViewsOnAppearWithTabBarController:topBar:mainView:bottomBar:useStatusBar:)
    class func resizeViewsOnAppear(withTabBarController tabBarController: UITabBarController?,
                                   topBar: UIView?,
                                   mainView: UIView?,
                                   bottomBar: UIView?,
                                   useStatusBar: Bool) {
        let screen = PSResizing.mainScreenBounds().size
        var topBarHeight: CGFloat = 0.0
        var bottomBarHeight: CGFloat = 0.0
        var viewHeight: CGFloat = 0.0
        var width: CGFloat = 0.0
        let tabBarHeight: CGFloat = (tabBarController != nil) ? tabBarController!.tabBar.frame.size.height : 0.0
        var redrawInNewFrames = false

        let interfaceOrientation = PSResizing.currentInterfaceOrientation()
        let statusBarSize = PSResizing.statusBarHeight()
        if interfaceOrientation == .landscapeLeft || interfaceOrientation == .landscapeRight {
            redrawInNewFrames = true
            bottomBarHeight = (bottomBar != nil) ? BOTTOM_BAR_LANDSCAPE_HEIGHT : 0.0
            width = screen.height
            topBarHeight = PSResizing.iPad() ? TOP_BAR_PORTRAIT_HEIGHT : TOP_BAR_LANDSCAPE_HEIGHT
            viewHeight = screen.width - topBarHeight - tabBarHeight - bottomBarHeight
            if useStatusBar {
                viewHeight -= statusBarSize
            }
        } else if interfaceOrientation == .portrait || interfaceOrientation == .portraitUpsideDown {
            redrawInNewFrames = true
            bottomBarHeight = (bottomBar != nil) ? BOTTOM_BAR_PORTRAIT_HEIGHT : 0.0
            width = screen.width
            topBarHeight = TOP_BAR_PORTRAIT_HEIGHT
            viewHeight = screen.height - topBarHeight - tabBarHeight - bottomBarHeight
            if useStatusBar {
                viewHeight -= statusBarSize
            }
        }
        if redrawInNewFrames {
            topBar?.frame = CGRect(x: 0.0, y: 0.0, width: width, height: topBarHeight)
            topBar?.layoutIfNeeded()
            topBar?.setNeedsDisplay()
            if let bottomBar = bottomBar {
                bottomBar.frame = CGRect(x: 0.0, y: (topBarHeight + viewHeight), width: width, height: bottomBarHeight)
                bottomBar.layoutIfNeeded()
                bottomBar.setNeedsDisplay()
            }
            mainView?.frame = CGRect(x: 0.0, y: topBarHeight, width: width, height: viewHeight)
        }
    }

    @objc(resizeViewsOnRotateWithTabBarController:topBar:mainView:fromOrientation:toOrientation:)
    class func resizeViewsOnRotate(withTabBarController tabBarController: UITabBarController?,
                                   topBar: UIView?,
                                   mainView: UIView?,
                                   fromOrientation fromInterfaceOrientation: UIInterfaceOrientation,
                                   toOrientation toInterfaceOrientation: UIInterfaceOrientation) {
        PSResizing.resizeViewsOnRotate(withTabBarController: tabBarController,
                                       topBar: topBar,
                                       mainView: mainView,
                                       bottomBar: nil,
                                       fromOrientation: fromInterfaceOrientation,
                                       toOrientation: toInterfaceOrientation)
    }

    @objc(resizeViewsOnRotateWithTabBarController:topBar:mainView:bottomBar:fromOrientation:toOrientation:)
    class func resizeViewsOnRotate(withTabBarController tabBarController: UITabBarController?,
                                   topBar: UIView?,
                                   mainView: UIView?,
                                   bottomBar: UIView?,
                                   fromOrientation fromInterfaceOrientation: UIInterfaceOrientation,
                                   toOrientation toInterfaceOrientation: UIInterfaceOrientation) {
        if PSResizing.iPad() {
            return
        }
        let screen = PSResizing.mainScreenBounds().size
        var topBarHeight: CGFloat = 0.0
        var bottomBarHeight: CGFloat = 0.0
        var viewHeight: CGFloat = 0.0
        var width: CGFloat = 0.0
        var bottomBarY: CGFloat = 0.0
        let tabBarHeight: CGFloat = (tabBarController != nil) ? tabBarController!.tabBar.frame.size.height : 0.0
        var redrawInNewFrames = false
        let statusBarSize = PSResizing.statusBarHeight()
        let toLandscape = (toInterfaceOrientation == .landscapeLeft || toInterfaceOrientation == .landscapeRight)
        let fromLandscape = (fromInterfaceOrientation == .landscapeLeft || fromInterfaceOrientation == .landscapeRight)
        let toPortrait = (toInterfaceOrientation == .portrait || toInterfaceOrientation == .portraitUpsideDown)
        let fromPortrait = (fromInterfaceOrientation == .portrait || fromInterfaceOrientation == .portraitUpsideDown)
        if toLandscape && !fromLandscape {
            width = screen.width
            bottomBarHeight = (bottomBar != nil) ? BOTTOM_BAR_LANDSCAPE_HEIGHT : 0.0
            bottomBarY = screen.height - statusBarSize - (bottomBarHeight / 2.0)
            topBarHeight = TOP_BAR_LANDSCAPE_HEIGHT
            viewHeight = screen.height - topBarHeight - tabBarHeight - bottomBarHeight - statusBarSize
            redrawInNewFrames = true
        } else if toPortrait && !fromPortrait {
            width = screen.height
            bottomBarHeight = (bottomBar != nil) ? BOTTOM_BAR_PORTRAIT_HEIGHT : 0.0
            bottomBarY = screen.width - statusBarSize - (bottomBarHeight / 2.0)
            topBarHeight = TOP_BAR_PORTRAIT_HEIGHT
            viewHeight = screen.width - topBarHeight - tabBarHeight - bottomBarHeight - statusBarSize
            redrawInNewFrames = true
        }
        if redrawInNewFrames {
            topBar?.frame = CGRect(x: 0.0, y: 0.0, width: width, height: topBarHeight)
            mainView?.center = CGPoint(x: (width / 2.0), y: ((viewHeight / 2.0) + topBarHeight))
            mainView?.bounds = CGRect(x: 0.0, y: 0.0, width: width, height: viewHeight)
            if let bottomBar = bottomBar {
                bottomBar.center = CGPoint(x: (width / 2.0), y: bottomBarY)
                bottomBar.bounds = CGRect(x: 0.0, y: 0.0, width: width, height: bottomBarHeight)
            }
        }
    }

    @objc(getOrientationRect:)
    class func getOrientationRect(_ interfaceOrientation: UIInterfaceOrientation) -> CGRect {
        let screen = PSResizing.mainScreenBounds().size
        return CGRect(x: 0.0, y: 0.0, width: screen.width, height: screen.height)
    }

    @objc(currentWindowScene)
    class func currentWindowScene() -> UIWindowScene? {
        for scene in UIApplication.shared.connectedScenes {
            if let windowScene = scene as? UIWindowScene {
                return windowScene
            }
        }
        return nil
    }

    @objc(keyWindow)
    class func keyWindow() -> UIWindow? {
        if let windowScene = PSResizing.currentWindowScene() {
            for window in windowScene.windows where window.isKeyWindow {
                return window
            }
            return windowScene.windows.first
        }
        return nil
    }

    @objc(mainScreenBounds)
    class func mainScreenBounds() -> CGRect {
        if let windowScene = PSResizing.currentWindowScene() {
            return windowScene.screen.bounds
        }
        // Fallback only when no window scene is connected yet (pre-scene launch).
        // UIScreen.main is deprecated in iOS 26 but has no context-free replacement
        // for bounds — a trait collection carries displayScale, not point bounds.
        return UIScreen.main.bounds
    }

    @objc(mainScreenScale)
    class func mainScreenScale() -> CGFloat {
        if let windowScene = PSResizing.currentWindowScene() {
            return windowScene.screen.scale
        }
        // UIScreen.main.scale is deprecated in iOS 26; the recommended replacement
        // is the current trait collection's displayScale.
        return UITraitCollection.current.displayScale
    }

    @objc(currentInterfaceOrientation)
    class func currentInterfaceOrientation() -> UIInterfaceOrientation {
        if let windowScene = PSResizing.currentWindowScene() {
            return windowScene.effectiveGeometry.interfaceOrientation
        }
        return .portrait
    }

    @objc(statusBarHeight)
    class func statusBarHeight() -> CGFloat {
        if let windowScene = PSResizing.currentWindowScene(),
           let statusBarManager = windowScene.statusBarManager {
            return statusBarManager.statusBarFrame.size.height
        }
        return 0.0
    }

    @objc(supportedInterfaceOrientations)
    class func supportedInterfaceOrientations() -> UIInterfaceOrientationMask {
        let rotationLockPosition = UserDefaults.standard.integer(forKey: Defaults.rotationLockPosition)
        switch rotationLockPosition {
        case kRotationEnabled:
            if PSResizing.iPad() {
                return .all
            } else {
                return .allButUpsideDown
            }
        case kRotationLockedInLandscape:
            return .landscape
        case kRotationLockedInPortrait:
            if PSResizing.iPad() {
                return [.portrait, .portraitUpsideDown]
            } else {
                return .portrait
            }
        default:
            return .all
        }
    }

    @objc(iPad)
    class func iPad() -> Bool {
        return UIDevice.current.userInterfaceIdiom != .phone
    }

    @objc(addSkipBackupAttributeToItemAtPath:)
    @discardableResult
    class func addSkipBackupAttribute(toItemAtPath path: String) -> Bool {
        // make sure we're not backing up this folder!
        dlog("Don't Backup:\n---\n\(path)\n---\n")

        let url = URL(fileURLWithPath: path, isDirectory: true)

        assert(FileManager.default.fileExists(atPath: url.path))

        do {
            var mutableURL = url
            var resourceValues = URLResourceValues()
            resourceValues.isExcludedFromBackup = true
            try mutableURL.setResourceValues(resourceValues)
            return true
        } catch {
            alog("Error excluding \(url.lastPathComponent) from backup \(error)")
            return false
        }
    }
}
